doc/packages/build-support.md

```diff
@@ -0,0 +1,92 @@
+# Build Support {#sec-build-support}
+
+## `pkgs.substitute` {#pkgs-substitute}
+
+`pkgs.substitute` is a wrapper around [the `substitute` function](#fun-substitute) provided by the
+stdenv. It replaces strings in `src` as specified by the `substitutions` argument.
+
+Whereas in a build script, one might call `substitute $infile $outfile --replace-fail @foo@
+${foopkg}/bin/foo`, one might accomplish this with `pkgs.substitute` as such:
+
+```nix
+{ substitute, foopkg }:
+substitute {
+  src = ./sourcefile.txt;
+  substitutions = [
+    "--replace"
+    "@foo@"
+    "${foopkg}/bin/foo"
+  ];
+}
+```
+
+## `pkgs.substituteAll` {#pkgs-substituteall}
+
+`pkgs.substituteAll` is a wrapper around [the `substituteAll` function](#fun-substituteAll) provided
+by the stdenv. It substitutes all instances of `@varName@` in `src` with the value of environment
+variable `varName`. As this uses the `substituteAll` function, its limitations also apply here,
+meaning that variables that you want to replace must:
+
+- start with a *lowercase* letter
+- only contain characters that are alphanumeric or `_`
+
+For example, if `say-goodbye.sh` contained the following:
+
+```
+#! @bash@/bin/bash
+
+echo @unchanged@
+@hello@/bin/hello --greeting @greeting@
+```
+
+the following derivation would make substitutions to `@bash@`, `@hello@`, and `@greeting@`:
+
+```nix
+{
+  substituteAll,
+  bash,
+  hello,
+}:
+substituteAll {
+  src = ./say-goodbye.sh;
+  inherit bash hello;
+  greeting = "goodbye";
+}
+```
+
+such that `$out` would result in something like the following:
+
+```
+#! /nix/store/s30jrpgav677fpc9yvkqsib70xfmx7xi-bash-5.2p26/bin/bash
+
+echo @unchanged@
+/nix/store/566f5isbvw014h7knmzmxa5l6hshx43k-hello-2.12.1/bin/hello --greeting goodbye
+```
+
+## `pkgs.substituteAllFiles` {#pkgs-substituteallfiles}
+
+Like `pkgs.substituteAll`, `pkgs.substituteAllFiles` is a wrapper around [the `substituteAll`
+function](#fun-substituteAll) provided by the stdenv. It funtions similarly by replacing `@varName@`
+with the value of the environment variable `varName`, however it expects `src` to be a directory
+and requires a `files` argument that specifies which files will be subject to replacements; only
+these files will be placed in `$out`.
+
+As it also uses the `substituteAll` function, it is subject to the same limitations on environment
+variables as discussed above in [pkgs.substituteAll](#pkgs-substituteall).
+
+For example, if the current directory contains `{foo,bar,baz}.txt` and the following `default.nix`
+
+```nix
+{ substituteAllFiles }:
+substituteAllFiles {
+  src = ./.;
+  files = [
+    "foo.txt"
+    "bar.txt"
+  ];
+  hello = "there";
+}
+```
+
+one would expect to find every instance of `@hello@` to be replaced with `there` in `$out/foo.txt`
```

Comment:
Can you use passive voice. Describe the actual output that the function produces.
e,g,

> every instance of `@hello@` will be replaced with `there` in `$out/foo.txt` and `$out/bar.txt`
> `baz.txt` wont be processed nor will it appear in `$out`.

Reply:
I think you mean active voice -- yes, happy to make that change.